File: backend/app/providers.py
"""Provider adapters — the only place third-party APIs are called.

Each adapter reads its key from settings (loaded from .env). If a key is
missing it returns status "not_configured" — NEVER a false "safe" result.
Network/auth failures return "unavailable" or "error". Keys are never logged.
"""

import asyncio
from urllib.parse import urlparse

import httpx

from .config import settings

_TIMEOUT = httpx.Timeout(15.0)


def _domain(url: str) -> str:
    net = urlparse(url).netloc or url
    return net.split("@")[-1].split(":")[0]


# ---------------------------------------------------------------------------
# VirusTotal — domain reputation for the URLs found in the email.
# ---------------------------------------------------------------------------
async def virustotal(client: httpx.AsyncClient, urls: list[str]) -> dict:
    key = settings.VIRUSTOTAL_API_KEY
    if not key:
        return {"provider": "VirusTotal", "status": "not_configured", "score": 0,
                "summary": "API key not configured."}
    domains = list(dict.fromkeys(_domain(u) for u in urls if _domain(u)))[:3]
    if not domains:
        return {"provider": "VirusTotal", "status": "ok", "score": 0, "summary": "No URLs to check."}
    try:
        malicious = 0
        for d in domains:
            r = await client.get(f"https://www.virustotal.com/api/v3/domains/{d}",
                                 headers={"x-apikey": key})
            if r.status_code == 200:
                stats = r.json().get("data", {}).get("attributes", {}).get("last_analysis_stats", {})
                malicious += int(stats.get("malicious", 0))
            elif r.status_code in (401, 403):
                return {"provider": "VirusTotal", "status": "unavailable", "score": 0,
                        "summary": "Access denied (check key/plan)."}
        score = min(40, malicious * 8)
        return {"provider": "VirusTotal", "status": "ok", "score": score,
                "summary": f"{malicious} malicious detection(s) across {len(domains)} domain(s)."}
    except Exception:
        return {"provider": "VirusTotal", "status": "unavailable", "score": 0,
                "summary": "Provider request failed."}


# ---------------------------------------------------------------------------
# Google Web Risk — malicious URL lookup.
# ---------------------------------------------------------------------------
async def web_risk(client: httpx.AsyncClient, urls: list[str]) -> dict:
    key = settings.GOOGLE_WEBRISK_API_KEY
    if not key:
        return {"provider": "Google Web Risk", "status": "not_configured", "score": 0,
                "summary": "API key not configured."}
    if not urls:
        return {"provider": "Google Web Risk", "status": "ok", "score": 0, "summary": "No URLs to check."}
    try:
        hits = 0
        for u in urls[:3]:
            r = await client.get(
                "https://webrisk.googleapis.com/v1/uris:search",
                params=[("key", key), ("uri", u),
                        ("threatTypes", "MALWARE"),
                        ("threatTypes", "SOCIAL_ENGINEERING")])
            if r.status_code == 200:
                if r.json().get("threat"):
                    hits += 1
            elif r.status_code in (401, 403):
                # e.g. billing disabled on the GCP project.
                return {"provider": "Google Web Risk", "status": "unavailable", "score": 0,
                        "summary": "Access denied (enable the API / billing)."}
        return {"provider": "Google Web Risk", "status": "ok", "score": 35 if hits else 0,
                "summary": f"{hits} URL(s) flagged." if hits else "No threats matched."}
    except Exception:
        return {"provider": "Google Web Risk", "status": "unavailable", "score": 0,
                "summary": "Provider request failed."}


# ---------------------------------------------------------------------------
# AbuseIPDB — sending-IP reputation.
# ---------------------------------------------------------------------------
async def abuseipdb(client: httpx.AsyncClient, ip: str | None) -> dict:
    key = settings.ABUSEIPDB_API_KEY
    if not key:
        return {"provider": "AbuseIPDB", "status": "not_configured", "score": 0,
                "summary": "API key not configured."}
    if not ip:
        return {"provider": "AbuseIPDB", "status": "ok", "score": 0, "summary": "No sending IP found."}
    try:
        r = await client.get("https://api.abuseipdb.com/api/v2/check",
                             params={"ipAddress": ip, "maxAgeInDays": 90},
                             headers={"Key": key, "Accept": "application/json"})
        if r.status_code in (401, 403):
            return {"provider": "AbuseIPDB", "status": "unavailable", "score": 0,
                    "summary": "Access denied (check key)."}
        data = r.json().get("data", {})
        confidence = int(data.get("abuseConfidenceScore", 0))
        return {"provider": "AbuseIPDB", "status": "ok", "score": round(confidence * 0.4),
                "summary": f"IP abuse confidence {confidence}%.",
                "abuse_score": confidence}
    except Exception:
        return {"provider": "AbuseIPDB", "status": "unavailable", "score": 0,
                "summary": "Provider request failed."}


# ---------------------------------------------------------------------------
# MaxMind GeoLite — IP geolocation / ASN. Needs account id + license key.
# ---------------------------------------------------------------------------
async def maxmind(client: httpx.AsyncClient, ip: str | None) -> dict:
    key = settings.MAXMIND_LICENSE_KEY
    account = settings.MAXMIND_ACCOUNT_ID
    if not key:
        return {"provider": "MaxMind GeoLite", "status": "not_configured", "score": 0,
                "summary": "License key not configured."}
    if not account:
        return {"provider": "MaxMind GeoLite", "status": "unavailable", "score": 0,
                "summary": "Set MAXMIND_ACCOUNT_ID to enable IP geolocation."}
    if not ip:
        return {"provider": "MaxMind GeoLite", "status": "ok", "score": 0, "summary": "No sending IP found."}
    try:
        r = await client.get(f"https://geolite.info/geoip/v2.1/city/{ip}",
                             auth=(account, key))
        if r.status_code in (401, 403):
            return {"provider": "MaxMind GeoLite", "status": "unavailable", "score": 0,
                    "summary": "Access denied (check account id / key)."}
        j = r.json()
        country = j.get("country", {}).get("iso_code")
        traits = j.get("traits", {})
        asn = traits.get("autonomous_system_number")
        isp = traits.get("isp") or traits.get("autonomous_system_organization")
        return {"provider": "MaxMind GeoLite", "status": "ok", "score": 0,
                "summary": f"Origin {country or '?'}.",
                "geo": {"country": country, "asn": f"AS{asn}" if asn else None, "isp": isp}}
    except Exception:
        return {"provider": "MaxMind GeoLite", "status": "unavailable", "score": 0,
                "summary": "Provider request failed."}


# ---------------------------------------------------------------------------
# MXToolbox — blacklist check for the sender domain.
# ---------------------------------------------------------------------------
async def mxtoolbox(client: httpx.AsyncClient, domain: str | None) -> dict:
    key = settings.MXTOOLBOX_API_KEY
    if not key:
        return {"provider": "MXToolbox", "status": "not_configured", "score": 0,
                "summary": "API key not configured."}
    if not domain:
        return {"provider": "MXToolbox", "status": "ok", "score": 0, "summary": "No domain to check."}
    try:
        r = await client.get(f"https://api.mxtoolbox.com/api/v1/lookup/blacklist/{domain}",
                             headers={"Authorization": key})
        if r.status_code in (401, 403):
            return {"provider": "MXToolbox", "status": "unavailable", "score": 0,
                    "summary": "Access denied (check key)."}
        failed = r.json().get("Failed", []) if r.headers.get("content-type", "").startswith("application/json") else []
        n = len(failed)
        return {"provider": "MXToolbox", "status": "ok", "score": min(20, n * 10),
                "summary": f"{n} blacklist hit(s)." if n else "Not blacklisted."}
    except Exception:
        return {"provider": "MXToolbox", "status": "unavailable", "score": 0,
                "summary": "Provider request failed."}


async def run_all(urls: list[str], sending_ip: str | None, sender_domain: str | None) -> list[dict]:
    """Run every provider concurrently and return their results in fixed order."""
    async with httpx.AsyncClient(timeout=_TIMEOUT) as client:
        results = await asyncio.gather(
            virustotal(client, urls),
            web_risk(client, urls),
            abuseipdb(client, sending_ip),
            maxmind(client, sending_ip),
            mxtoolbox(client, sender_domain),
        )
    return list(results)

File: supabase/functions/analyze-email/providers.ts
// ============================================================================
// Provider adapters.
// ----------------------------------------------------------------------------
// Each adapter reads its API key from an ENVIRONMENT VARIABLE. If the key is
// missing, it returns status "not_configured" — NEVER a false "safe" result.
//
// SECURITY:
//   - No key is hardcoded here. Configure real secrets later from VS Code:
//       supabase secrets set VIRUSTOTAL_API_KEY=xxxx
//       supabase secrets set GOOGLE_WEBRISK_API_KEY=xxxx
//       supabase secrets set ABUSEIPDB_API_KEY=xxxx
//       supabase secrets set MXTOOLBOX_API_KEY=xxxx
//       supabase secrets set MAXMIND_LICENSE_KEY=xxxx
//   - Never log the key or the raw email body.
//
// The network calls below are written as commented templates. Until you enable
// them, each adapter reports "not_configured" (missing key) so the report is
// honest about what was and wasn't checked.
// ============================================================================

export type ProviderStatus = "ok" | "not_configured" | "unavailable" | "error";

export interface ProviderResult {
  provider: string;
  status: ProviderStatus;
  score: number; // 0..100 contribution
  summary: string;
}

/** Helper: read an env var without ever logging its value. */
function key(name: string): string | undefined {
  const v = Deno.env.get(name);
  return v && v.trim() ? v : undefined;
}

// ---------------------------------------------------------------------------
// VirusTotal — URL/domain reputation.
// ---------------------------------------------------------------------------
export async function checkVirusTotal(_urls: string[]): Promise<ProviderResult> {
  const k = key("VIRUSTOTAL_API_KEY");
  if (!k) {
    return { provider: "VirusTotal", status: "not_configured", score: 0,
             summary: "API key not configured." };
  }
  try {
    // const res = await fetch("https://www.virustotal.com/api/v3/urls", {
    //   method: "POST",
    //   headers: { "x-apikey": k, "content-type": "application/x-www-form-urlencoded" },
    //   body: new URLSearchParams({ url: _urls[0] ?? "" }),
    // });
    // ...parse res, map malicious/suspicious counts to a 0..100 score...
    return { provider: "VirusTotal", status: "ok", score: 0,
             summary: "No malicious detections." };
  } catch {
    return { provider: "VirusTotal", status: "unavailable", score: 0,
             summary: "Provider request failed." };
  }
}

// ---------------------------------------------------------------------------
// Google Web Risk — malicious URL lookup.
// ---------------------------------------------------------------------------
export async function checkWebRisk(_urls: string[]): Promise<ProviderResult> {
  const k = key("GOOGLE_WEBRISK_API_KEY");
  if (!k) {
    return { provider: "Google Web Risk", status: "not_configured", score: 0,
             summary: "API key not configured." };
  }
  try {
    // const res = await fetch(
    //   `https://webrisk.googleapis.com/v1/uris:search?key=${k}&uri=${encodeURIComponent(_urls[0] ?? "")}&threatTypes=MALWARE&threatTypes=SOCIAL_ENGINEERING`,
    // );
    return { provider: "Google Web Risk", status: "ok", score: 0,
             summary: "No threats matched." };
  } catch {
    return { provider: "Google Web Risk", status: "unavailable", score: 0,
             summary: "Provider request failed." };
  }
}

// ---------------------------------------------------------------------------
// AbuseIPDB — sending-IP reputation.
// ---------------------------------------------------------------------------
export async function checkAbuseIPDB(ip: string | null): Promise<ProviderResult> {
  const k = key("ABUSEIPDB_API_KEY");
  if (!k) {
    return { provider: "AbuseIPDB", status: "not_configured", score: 0,
             summary: "API key not configured." };
  }
  if (!ip) {
    return { provider: "AbuseIPDB", status: "ok", score: 0, summary: "No sending IP found." };
  }
  try {
    // const res = await fetch(
    //   `https://api.abuseipdb.com/api/v2/check?ipAddress=${ip}&maxAgeInDays=90`,
    //   { headers: { Key: k, Accept: "application/json" } },
    // );
    // const data = await res.json();
    // const confidence = data?.data?.abuseConfidenceScore ?? 0;
    // return { provider: "AbuseIPDB", status: "ok", score: Math.round(confidence * 0.4),
    //          summary: `IP abuse confidence ${confidence}%.` };
    return { provider: "AbuseIPDB", status: "ok", score: 0, summary: "Lookup complete." };
  } catch {
    return { provider: "AbuseIPDB", status: "unavailable", score: 0,
             summary: "Provider request failed." };
  }
}

// ---------------------------------------------------------------------------
// MaxMind GeoLite — IP geolocation / ASN.
// ---------------------------------------------------------------------------
export async function checkMaxMind(ip: string | null): Promise<ProviderResult> {
  const k = key("MAXMIND_LICENSE_KEY");
  if (!k) {
    return { provider: "MaxMind GeoLite", status: "not_configured", score: 0,
             summary: "License key not configured." };
  }
  if (!ip) {
    return { provider: "MaxMind GeoLite", status: "ok", score: 0, summary: "No sending IP found." };
  }
  try {
    // Use the GeoIP2 web service or a local GeoLite2 DB. Geolocation itself is
    // not a threat signal, so score stays 0; it enriches the report context.
    return { provider: "MaxMind GeoLite", status: "ok", score: 0, summary: "Geolocation resolved." };
  } catch {
    return { provider: "MaxMind GeoLite", status: "unavailable", score: 0,
             summary: "Provider request failed." };
  }
}

// ---------------------------------------------------------------------------
// MXToolbox — blacklist / mail-server reputation.
// ---------------------------------------------------------------------------
export async function checkMXToolbox(domain: string | null): Promise<ProviderResult> {
  const k = key("MXTOOLBOX_API_KEY");
  if (!k) {
    return { provider: "MXToolbox", status: "not_configured", score: 0,
             summary: "API key not configured." };
  }
  if (!domain) {
    return { provider: "MXToolbox", status: "ok", score: 0, summary: "No domain to check." };
  }
  try {
    // const res = await fetch(`https://api.mxtoolbox.com/api/v1/lookup/blacklist/${domain}`,
    //   { headers: { Authorization: k } });
    return { provider: "MXToolbox", status: "ok", score: 0, summary: "Not blacklisted." };
  } catch {
    return { provider: "MXToolbox", status: "unavailable", score: 0,
             summary: "Provider request failed." };
  }
}
